File: graphql-dgs-client/src/main/java/com/netflix/graphql/dgs/client/DgsRestClientGraphQLClient.java
/*
 * Copyright 2025 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.netflix.graphql.dgs.client;

import com.netflix.graphql.dgs.json.DgsJsonMapper;
import org.intellij.lang.annotations.Language;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestClient;

import java.util.Map;
import java.util.function.Consumer;

/**
 * RestClient-based blocking DGS client.
 * The no-arg convenience constructor uses Jackson 3 under the hood. Callers on Jackson 2
 * must pass {@link Jackson2DgsJsonMapperAdapter} explicitly.
 */
public class DgsRestClientGraphQLClient implements DgsGraphQLClient {
    private final RestClient restClient;
    private final Consumer<HttpHeaders> headersConsumer;
    private final DgsJsonMapper mapper;

    public DgsRestClientGraphQLClient(
            RestClient restClient, Consumer<HttpHeaders> headersConsumer, DgsJsonMapper mapper) {
        this.restClient = restClient;
        this.headersConsumer = headersConsumer;
        this.mapper = mapper;
    }

    public DgsRestClientGraphQLClient(RestClient restClient) {
        this(restClient, headers -> { });
    }

    public DgsRestClientGraphQLClient(RestClient restClient, Consumer<HttpHeaders> headersConsumer) {
        this(restClient, headersConsumer, Jackson3DgsJsonMapperAdapter.defaultMapper());
    }

    public DgsRestClientGraphQLClient(RestClient restClient, DgsGraphQLRequestOptions options) {
        this(restClient, headers -> { }, Jackson3DgsJsonMapperAdapter.fromOptions(options));
    }

    @Override
    public DgsGraphQLResponse executeQuery(@Language("graphql") String query) {
        return executeQuery(query, Map.of(), null);
    }

    @Override
    public DgsGraphQLResponse executeQuery(@Language("graphql") String query, Map<String, Object> variables) {
        return executeQuery(query, variables, null);
    }

    @Override
    public DgsGraphQLResponse executeQuery(
            @Language("graphql") String query, Map<String, Object> variables, String operationName) {
        String serializedRequest = mapper.writeValueAsString(GraphQLClients.toRequestMap(query, operationName, variables));

        ResponseEntity<String> responseEntity =
                restClient
                        .post()
                        .headers(headers -> GraphQLClients.defaultHeaders.forEach(headers::addAll))
                        .headers(this.headersConsumer)
                        .body(serializedRequest)
                        .retrieve()
                        .toEntity(String.class);

        if (!responseEntity.getStatusCode().is2xxSuccessful()) {
            throw new GraphQLClientException(
                    responseEntity.getStatusCode().value(),
                    "",
                    responseEntity.getBody() != null ? responseEntity.getBody() : "",
                    serializedRequest);
        }

        return new DefaultDgsGraphQLResponse(
                responseEntity.getBody() != null ? responseEntity.getBody() : "",
                HttpHeaderUtils.toMap(responseEntity.getHeaders()),
                mapper);
    }
}

File: graphql-dgs-client/src/main/kotlin/com/netflix/graphql/dgs/client/DgsRestClientGraphQLClient.kt
/*
 * Copyright 2026 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 */

package com.netflix.graphql.dgs.client

import com.netflix.graphql.dgs.json.DgsJsonMapper
import org.intellij.lang.annotations.Language
import org.springframework.http.HttpHeaders
import org.springframework.web.client.RestClient
import org.springframework.web.client.toEntity
import java.util.function.Consumer

/**
 * RestClient-based blocking DGS client.
 *
 * The no-arg convenience constructor uses Jackson 3 under the hood. Callers on Jackson 2
 * must pass [Jackson2DgsJsonMapperAdapter] explicitly.
 */
class DgsRestClientGraphQLClient(
    private val restClient: RestClient,
    private val headersConsumer: Consumer<HttpHeaders>,
    private val mapper: DgsJsonMapper,
) : DgsGraphQLClient {
    constructor(restClient: RestClient) : this(restClient, Consumer { })

    constructor(restClient: RestClient, headersConsumer: Consumer<HttpHeaders>) :
        this(restClient, headersConsumer, Jackson2DgsJsonMapperAdapter.default())

    constructor(restClient: RestClient, options: DgsGraphQLRequestOptions) :
        this(restClient, Consumer { }, Jackson2DgsJsonMapperAdapter.fromOptions(options))

    override fun executeQuery(
        @Language("graphql") query: String,
    ): DgsGraphQLResponse = executeQuery(query, emptyMap(), null)

    override fun executeQuery(
        @Language("graphql") query: String,
        variables: Map<String, Any>,
    ): DgsGraphQLResponse = executeQuery(query, variables, null)

    override fun executeQuery(
        @Language("graphql") query: String,
        variables: Map<String, Any>,
        operationName: String?,
    ): DgsGraphQLResponse {
        val serializedRequest =
            mapper.writeValueAsString(
                GraphQLClients.toRequestMap(query = query, operationName = operationName, variables = variables),
            )

        val responseEntity =
            restClient
                .post()
                .headers { headers ->
                    GraphQLClients.defaultHeaders.forEach { (key, values) ->
                        headers.addAll(key, values)
                    }
                }.headers(this.headersConsumer)
                .body(serializedRequest)
                .retrieve()
                .toEntity<String>()

        if (!responseEntity.statusCode.is2xxSuccessful) {
            throw GraphQLClientException(
                statusCode = responseEntity.statusCode.value(),
                url = "",
                response = responseEntity.body ?: "",
                request = serializedRequest,
            )
        }

        return DefaultDgsGraphQLResponse(json = responseEntity.body ?: "", headers = responseEntity.headers.toMap(), mapper)
    }
}

private fun HttpHeaders.toMap(): Map<String, List<String>> {
    val result = mutableMapOf<String, List<String>>()
    this.forEach { key, values -> result[key] = values }
    return result
}
